### PR TITLE
Week 1 Assignment

### PR DESCRIPTION
**Pull request by jeelion22**
initial completion

### CODE REVIEW

**Review by josef-o — CHANGES_REQUESTED** [review]
### Ex 1

You have to `return` the value instead of printing it.  
You have to use a `ValueError` as the exception type instead of a general `Exception` that would catch anything (and don't have to print out the traceback or exception type).

### Ex 2

Again in the `try` block catch the specific exception to be handled.  
The `else` part would be better to have inside the `try` block.  
You aren't returning the contents only printing them but maybe that's fine, it's not completely clear in the question.

### Ex 3

Here you aren't doing what is asked, in the example you gave the result should be `[1.2]`.  
By the way it's really cool that you are using `map` that's quite advanced (but have to be careful also).  
Again `except Exception` is a bad practice.

### Ex 4

This one is pretty clean, again you are printing the answers instead of returning them but maybe that's fine because it's cleaner that way.  
Here also change the exception type.  
The `continue` is not needed as there is nothing after it to skip.

### Ex 5

Exception type.  
Good idea with the `None` but there's a bug if the first element in the list is not an integer.  
```
largest_int(['1.7', 1])
> TypeError: 1 is a string. 
> Only integers list allowed.
```

**Comment by jeelion22** [comment]
😲 wow awesome review, I will look into this.

File: error_handling.ipynb [hunk 1]
```diff
@@ -28,7 +28,10 @@
     "    try:\n",
     "        map_obj = map(int, int_list)\n",
     "        int_list = list(map_obj)\n",
+    "        \n",
     "        sum = 0\n",
+    "        \n",
+    "        # sum the items in the intlist\n",
     "\n",
     "        for i in range(len(int_list)):\n",
```
Comment (rzmobiledev) [inline error_handling.ipynb:36]: what about removing this loop. just :

`data = sum(int_list)'
`return data`

File: error_handling.ipynb [hunk 2]
```diff
@@ -106,16 +85,15 @@
     "\n",
     "    else:\n",
```
Comment (rzmobiledev) [inline error_handling.ipynb:86]: Where is `if` statement` here? i see you put `else` condition

File: error_handling.ipynb [hunk 3]
```diff
@@ -154,6 +181,59 @@
...
+    "           \n",
+    "        for i in lst:\n",
+    "            if largest is None or i > largest :\n",
+    "                largest = i\n",
+    "        \n",
+    "    except Exception as e:\n",
+    "        print('TypeError: {} is a string. '.format(i))\n",
+    "        print('Only integers list allowed.')\n",
+    "                \n",
+    "    else :\n",
```
Comment (rzmobiledev) [inline error_handling.ipynb:215]: just remove `print statement` from `else block`. You are safe to delete `else statement` here

File: error_handling.ipynb (hunk 1 above)
Comment (jeelion22) [reply error_handling.ipynb:36]: yes, agree that

File: error_handling.ipynb (hunk 3 above)
Comment (jeelion22) [reply error_handling.ipynb]: done

**Review by josef-o — CHANGES_REQUESTED** [review]
### Ex 1

Great that you are using `ValueError` exception but why the `TypeError` one?  
In the `return` part you are casting the `int_list` first to `float` and then to `int` which seems redundant.  
Also casiting the inner `map` to `list` isn't really needed because the `sum` doesn't mind it.

*(by the way using `map` is quite powerful and makes for short programs but sometimes they become overly complicated or hard to understand so be careful with that)*

### Ex 2

This is one is better now.  
The `file_path` being an optional argument now makes more sense as it wasn't asked for.

### Ex 3

Be careful with using `float(i)` as the condition because of numbers truthy values in python.  

Bug: 
```
float_convertables([0, "0", 1.2, '2', 'a', 3, 'float']) 
>>> ['1.2', '2', '3']
```

### Ex 4

Much better now.  
Small suggestion for making the `print` cleaner is using double quotes for the string so that way you don't need to escape the single quotes inside: `print("The key '{}' is not found".format(key))`

### Ex 5

Same thing as **Ex 1** casting to `float` then `list` then `int` is not needed.

**Comment by josef-o** [comment]
### Ex 1

I get your motivation on casting to float first but it would give a wrong result as 1.5 to int would be 1, which is different than 1.5.

But more to the point it's doing things the function isn't supposed to do (as was asked for), same with adding extra exceptions or a general `except Exception`, if the wrong data structure is being passed to the function it is okay that it breaks.

### Ex 3

You haven't really fixed it, you attacked one of the symptoms rather than the cause (which you need to understand first).
```
float_convertables(["-0", '0.0', 1.2, '2', 'a', 3, 'float'])
>>> ['1.2', '2', '3']
```

### Ex5 

Ref.to **Ex1** review reply above :smiley:

**Comment by jeelion22** [comment]
Changes are done

**Review by josef-o — APPROVED** [review]
Looks good now :+1:

**Comment by jeelion22** [comment]
shall I merge with the main branch?

**Comment by josef-o** [comment]
I have no further suggestions but that's up to you of course if you feel you are done with this assignment.

**Comment by jeelion22** [comment]
Okay, and thanks. Now I close the pr #3, and I try further trim them. once
done I make it

**Comment by jeelion22** [comment]
shall I close it or leave it?

**Comment by josef-o** [comment]
To be clear, closing means you are not accepting the changes from this branch which I don't think is what you want to do.

**Comment by jeelion22** [comment]
No, I'm not.

**Comment by jeelion22** [comment]
I created another branch for week2 assignment. So I thought, i can make
multi pr .

**Comment by jeelion22** [comment]
Is it bad practice?

**Comment by josef-o** [comment]
I think I get what you did, correct me if I'm wrong but you created a branch called `assignment_week_1` with the same commits as this one so you can merge from there and the branch you merge from has a meaningful name.  

It's a good practice to name your branches that way yes, especially when there are many PRs open at the same time. It's also acceptable to work in a single branch (in this case `development`) and merge from there everytime. 

The thing is that the branch itself doesn't matter that much and it's a good practice to delete it after it is merged. Then the next pr can be made from a branch with the same name **but here's the important part:** even though it has the same name it's a different branch because you create it from the updated `main` branch with the newest code (its parent commit is newer).

With all that said, more important than the branches are the PRs themselves with all the explanations and discussions that document well the history and reason for all the changes in the code in a focused manner.

Here I recommend you keep this PR and merge it. But also that you change it's name and description at the top to better reflect its purpose.

You can rename it from "Development" to "Week 1 assignments" and in the description add something like the names of the files that are worked on, or the difficulty of each one like what Arsalan put in his email.

And it is better that you create the branch and PR for the next assignment **after** you merge this one to try to avoid merge conflicts in the future or having to merge main into the PR which can make for a convoluted history.

Anyways, sorry for the wall of text :sweat_smile:

**Comment by jeelion22** [comment]
I got it, thanks! You well said, don't worry about it. It's a good comment
and will help understand git more. So What I do now is, merge this
branch(development) with the main, then rename it "week_1_assignment", and
add file names in the description that I worked in the development branch.
My question is once done, will the pr be automatically closed after
renaming the existing branch? because I tried to rename it last time
without merging with the main, pr #3 closed automatically, so I reopened pr
#3 again. After merging and renaming, will the pr open? I have another
branch named assignment_week_2 where I have worked files, shall I raise pr
for them from there

**Comment by jeelion22** [comment]
once, could you come to meet to look at my branches?

**Pull request merged**